### PR TITLE
Files Changed Overview (FCO)

### PR DESCRIPTION
**Pull request by playcations**
### Related GitHub Issue

Closes: # [4454](https://github.com/RooCodeInc/Roo-Code/issues/4454)

### Description

This adds a collapsable window in chat to show files that have been modified by agents. 

Features include accepting changes, diffing and rejecting file changes outright. 
This prevents you from needing to keep git open or going back to a previous checkpoint after every big modifications to the codebase. I definitely had times where some good changes have happened but then the agent decides theya ren't done! You can easily reject some of the changes instead of all of them. If you don't like this feature, you can also turn it off in the settings menu. 

Checkpoints system did need modifications to get this to work properly. You can see what was done there and let me know what you think. 
ShadowCheckpoints are now stored in a .roo folder in root. This prevents the need to search for all git files and turn off the system if some are found not in root. This was a real bummer when I went to use roo on larger codebases that have .gits throughout. 

### Test Procedure
open chat and ask it to modify some files, make new files, etc. accept or reject the changes
open settings menu and under interfaces, turn off the feature, go through the steps again 
turn on the feature on last time and go through the steps again. 

make sure checkpoints still work. 

### Pre-Submission Checklist

- [X] **Issue Linked**: This PR is linked to an approved GitHub Issue (see "Related GitHub Issue" above).
- [X] **Scope**: My changes are focused on the linked issue (one major feature/fix per PR).
- [X] **Self-Review**: I have performed a thorough self-review of my code.
- [X] **Testing**: New and/or updated tests have been added to cover my changes (if applicable).
- [X] **Documentation Impact**: I have considered if my changes require documentation updates (see "Documentation Updates" section below).
- [X] **Contribution Guidelines**: I have read and agree to the [Contributor Guidelines](/CONTRIBUTING.md).

#### Before:
image

#### After Checkpoint is hit: 
image
image

#### Settings option: 
image

### Documentation Updates

You can add a line like, "After every checkpoint, Files changed will appear in chat. You have the option to accept or reject them on a file by file basis. to turn this feature off, find it in the settings menu" but I don't think it makes much to consider it significant. 

### Additional Notes

I consolidated the feature, made a new branch, pushed just feature changes, and squashed it into a single commit to only contain changes made that involve this feature. 

@hannesrudolph I was not sure what you meant by breaking it down by PR as it is one PR. If you have more information, I can get more descriptive. please let me know. If this is due to ShadowCheckpoints changes and there is an issue for it, I can add it. 

----

- The nested git repository detection test was failing because FCO integration, changed the shadow git implementation to use .roo directory approach, eliminating the need for nested git repository detection.
- add missing UI components and webviewReady message
- Import and render FilesChangedOverview component in ChatView
- Add webviewReady message to useMount hook for FCO initialization
- Add FCO message handling delegation while preserving existing routing (INTEGRATION)
- Add FileChangeManager integration for file change tracking (CORE FEATURE)
- Add globalFileChangeManager property and disposal cleanup (LIFECYCLE)
- Add FCO state properties to provider state (STATE ENHANCEMENT)
- Add FileChangeManager accessor methods for FCO components (API)
- Add type annotations for task creation callbacks (TYPE SAFETY)
- Add filesChangedEnabled message case with state update (NEW MESSAGE HANDLING)
- Add filesChangedEnabled property to state object construction (STATE INTEGRATION)
- Enhance error handling for API configuration operations (ERROR HANDLING)
- Preserve all existing message handling functionality
- Add FCO state properties to ExtensionStateContext interface (ENHANCEMENT)
- Add filesChanged message handling for real-time updates (NEW CAPABILITY)
- Add FCO state update functions for component interaction (STATE MANAGEMENT)
- Add currentFileChangeset state for file change tracking (CORE FEATURE)
- Initialize filesChangedEnabled in default state (INITIALIZATION)
- Add 9 FCO webview message types to existing union (ADDITIVE)
- Add 4 FCO extension message types (ADDITIVE)
- Add Terminal interface for enhanced functionality (NEW)
- Add optional properties for FCO file operations (ENHANCEMENT)
- Add filesChangedEnabled state property (STATE ENHANCEMENT)
- Preserve all existing message types and backward compatibility
- Add FCO component mocks (DecorationController, DiffViewProvider) (TEST INFRASTRUCTURE)
- Enhance VSCode mocks with createTextEditorDecorationType, Range, Position (API MOCKS)
- Add filesChangedEnabled to test state expectations (STATE TESTING)
- Preserve all existing test functionality and coverage
- Export FCO types from packages/types (MINIMAL ADDITION)
- CheckpointDiff added change type classification (TYPE ENHANCEMENT)
- Update checkpoint event interface fto no longer check for nested gits and instead use .roo for all checkpoints (EVENT ENHANCEMENT)
- Preserve all main branch type system improvements
- Add FileChange and FileChangeset types (NEW)
- Add FileChangeManager for state management (NEW)
- Add FCOMessageHandler for webview communication (NEW)
- Add FilesChangedOverview React components (NEW)
- Add comprehensive FCO test coverage (NEW)
- Add full internationalization support (18 languages) (NEW)
- UI section and FilesChangedEnabled flag added to Settings
- testing added to UI section
- Add optional `uris` parameter to WebviewMessage type
- Update frontend to send visible file URIs with reject all message
- Modify backend to filter files based on provided URIs
- Maintain backward compatibility (defaults to all files if no URIs)
- Add comprehensive test coverage including scoped rejection scenarios
- Fix ShadowCheckpointService test - comment out obsolete nested git check

### Get in Touch

Playcations on Discord

### CODE REVIEW

**Comment by adamhill** [comment]
Wow. @playcations This is NICE!!!

**Comment by playcations** [comment]
@adamhill I forgot the language files and am failing the cloud tests.. is that okay? I know my edits changed the shadow checkpoint system but running tests locally didn't test this.

**Comment by playcations** [comment]
I see I failed cloudServices and I forgot the language changes for settings... I can work on adding those later today. My mistake. I can close this and make another pr if need be.

**Comment by playcations** [comment]
updated:
- updated language files for settings to include FilesChangedOverview.
- made Files changed overview match closer to Todo list since that is the theming now. (nice and slim)
- small bug fix and added test procedure for the added todo feature and working with it properly.

**Comment by playcations** [comment]
Fco edge case fixes and tests (final push until review)
- folders were getting diffed, added a check to prevent this.
- .roo now excluded from checkpoint diffs
- diffs for FCO now only check line changes vs full diff changes. This is a preventative measure due to search and replace calls rewriting the entire file.
- tests for these edge cases added.

**Comment by mkdir700** [comment]
That's great

File: src/shared/ExtensionMessage.ts [hunk 1]
```diff
@@ -105,6 +107,10 @@ export interface ExtensionMessage {
 		| "shareTaskSuccess"
 		| "codeIndexSettingsSaved"
 		| "codeIndexSecretStatus"
+		| "filesChanged"
+		| "checkpoint_created"
+		| "checkpoint_restored"
```
Comment (mkdir700) [inline src/shared/ExtensionMessage.ts]: Perhaps we can maintain stylistic consistency?

**Review by mkdir700 — CHANGES_REQUESTED** [review]

File: webview-ui/src/components/file-changes/FilesChangedOverview.tsx [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+			// Guard against null/undefined/malformed messages
+			if (!message || typeof message !== "object" || !message.type) {
+				console.debug("[FCO] Ignoring malformed message:", message)
+				return
+			}
+
+			switch (message.type) {
+				case "filesChanged":
+					if (message.filesChanged) {
+						console.log("[FCO] Received filesChanged message:", message.filesChanged)
```
Comment (daniel-lxs) [inline webview-ui/src/components/file-changes/FilesChangedOverview.tsx]: I noticed some `console.log` and `console.debug` statements here. Are these intended for debugging and should they be removed before merging?

File: webview-ui/src/components/file-changes/FilesChangedOverview.tsx [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+	// Don't render if the feature is disabled or no changes to show
+	if (!filesChangedEnabled || !changeset || files.length === 0) {
+		return null
+	}
+
+	return (
+		<div
+			className="files-changed-overview"
+			data-testid="files-changed-overview"
+			style={{
```
Comment (daniel-lxs) [inline webview-ui/src/components/file-changes/FilesChangedOverview.tsx]: I see some inline styles here. To keep the styling consistent across the application, would it be better to move these to CSS classes?

File: webview-ui/src/components/file-changes/FilesChangedOverview.tsx [hunk 4]
```diff
@@ -0,0 +1,516 @@
+import React from "react"
+import { FileChangeset, FileChange } from "@roo-code/types"
+import { useTranslation } from "react-i18next"
+import { useExtensionState } from "@/context/ExtensionStateContext"
+import { vscode } from "@/utils/vscode"
+
+// eslint-disable-next-line @typescript-eslint/no-empty-object-type
+interface FilesChangedOverviewProps {}
+
+interface _CheckpointEventData {
```
Comment (daniel-lxs) [inline webview-ui/src/components/file-changes/FilesChangedOverview.tsx]: It looks like this interface is unused. Can it be removed?

File: webview-ui/src/components/file-changes/FilesChangedOverview.tsx [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+		// Backend will send updated filesChanged message with filtered results
+	}, [])
+
+	const handleRejectAll = React.useCallback(() => {
+		const visibleUris = files.map((file) => file.uri)
+		vscode.postMessage({ type: "rejectAllFileChanges", uris: visibleUris })
+		// Backend will send updated filesChanged message with filtered results
+	}, [files])
+
+	const handleWithDebounce = React.useCallback(
```
Comment (daniel-lxs) [inline webview-ui/src/components/file-changes/FilesChangedOverview.tsx]: The debouncing logic is implemented directly in this component. Have you considered extracting it into a reusable custom hook? This could help improve separation of concerns.

**Review by daniel-lxs — CHANGES_REQUESTED** [review]
Thank you for taking the time to implement this feature, I feel like it's a nice idea however I do think it would be better to move this to the experimental settings section, and that means removing the new UI section.

I also left a couple of minor points that I think are worth solving.

Let me know what you think!

**Comment by hannesrudolph** [comment]
@roomote-agent fix the merge conflicts using pr-fixer mode

**Comment by hannesrudolph** [comment]
@roomote-agent fix the merge conflicts using pr-fixer mode

**Comment by playcations** [comment]
@daniel-lxs I was actively working on changes since your push. I hope my recent perge push didn't breach anything. Still getting used to git. 

edit: removed the merge I made yesterday. I have one more change I need to make and then I think it will be ready for review again.

**Comment by playcations** [comment]
Just got home today, I'll be working on the final touches tomorrow. Thank you for your patience.

**Comment by playcations** [comment]
Just made a final push. FCO is moved, cleaned up code. Added more test cases found during cleanup. debouncing logic separated, Inline styling moved, stylistic consistency updated for @mkdir700 's concerns. hopefully this covers everything.

**Comment by playcations** [comment]
Hey, any update? Is there something I'm forgetting to do?

**Comment by daniel-lxs** [comment]
Hey @playcations, sorry for the delay. I'll be reviewing this again soon. Thank you!

**Comment by playcations** [comment]
@daniel-lxs No worries, I was just worried I did something wrong. Excuse the spam notifications if you got them.

**Comment by hannesrudolph** [comment]
Status update: All requested changes have been addressed and CI is green across all platforms.

Key fixes:
- Moved FCO toggle to Experimental and removed dedicated UI section:
  • [ExperimentalSettings.tsx](webview-ui/src/components/settings/ExperimentalSettings.tsx)
  • [SettingsView.tsx](webview-ui/src/components/settings/SettingsView.tsx)
  • [UISettings.tsx](webview-ui/src/components/settings/UISettings.tsx)
  • [UISettings.spec.tsx](webview-ui/src/components/settings/__tests__/UISettings.spec.tsx)
- Extracted debouncing to reusable hook and integrated into FCO:
  • [useDebouncedAction.ts](webview-ui/src/components/ui/hooks/useDebouncedAction.ts)
  • [FilesChangedOverview.tsx](webview-ui/src/components/file-changes/FilesChangedOverview.tsx)
- Replaced inline styles with utility classes in FCO where possible:
  • [FilesChangedOverview.tsx](webview-ui/src/components/file-changes/FilesChangedOverview.tsx)
- Stabilized checkpoint diff path handling to avoid CI path mismatches:
  • [ShadowCheckpointService.initShadowGit()](src/services/checkpoints/ShadowCheckpointService.ts:74)
- FCO message handling verified and refined:
  • [FCOMessageHandler.ts](src/services/file-changes/FCOMessageHandler.ts)
- Stylistic consistency cleanup in shared message types:
  • Deduped union entries and added alias for payload naming consistency:
    - [WebviewMessage.type](src/shared/WebviewMessage.ts:25)
    - [WebviewMessagePayload alias](src/shared/WebviewMessage.ts:342)
  • Clarified Extension → Webview message comment header:
    - [ExtensionMessage](src/shared/ExtensionMessage.ts:60)
- i18n: Added/standardized required keys across locales (settings + file-changes), including Indonesian terms.

CI results:
- compile, knip, translations, CodeQL, linux/windows unit tests all passing.

Requesting re-review.
@daniel-lxs @mkdir700 please take another look. If anything else is needed, I can adjust quickly.

**Comment by playcations** [comment]
Hey, I am sorry to do this, and this is completely my bad. I did not push my last commit and will need to do that as design was changed and I found a bug that i did not push. I am very sorry about this. This is completely my fault.

**Review by playcations — COMMENTED** [review]
There are a couple things that seemed to me missed that I am pushing.

**Review by playcations — COMMENTED** [review]
Chat only changes made by llms are now tracked*

File: src/core/checkpoints/index.ts [hunk 6]
```diff
@@ -26,6 +29,12 @@ export async function getCheckpointService(
 	if (cline.checkpointService) {
 		return cline.checkpointService
 	}
+	console.log(
+		`[DEBUG] getCheckpointService called for task ${cline.taskId}. Service exists: ${!!cline.checkpointService}`,
+	)
+	if (!cline.enableCheckpoints) {
```
Comment (NaccOll) [inline src/core/checkpoints/index.ts]: duplicate with 26-28

File: src/core/checkpoints/index.ts [hunk 7]
```diff
@@ -122,25 +131,211 @@ async function checkGitInstallation(
...
+						}
+					}
+				} catch (error) {
+					log(`[Task#getCheckpointService] Failed to update FileChangeManager baseline: ${error}`)
+					// Don't throw - allow checkpoint service to continue initializing
+				}
+
+				if (isCheckpointNeeded) {
+					log("[Task#getCheckpointService] no checkpoints found, saving initial checkpoint")
+					checkpointSave(cline, true)
```
Comment (NaccOll) [inline src/core/checkpoints/index.ts]: Currently, Disabled the checkpoint creation during initialization and postponed it until the first time a file is edited. Is this change necessary?

File: src/core/checkpoints/index.ts [hunk 8]
```diff
@@ -161,20 +356,104 @@ async function checkGitInstallation(
 	}
 }
 
-export async function checkpointSave(cline: Task, force = false) {
+export async function getInitializedCheckpointService(
```
Comment (NaccOll) [inline src/core/checkpoints/index.ts:352]: This method has been removed. CheckpointService is now obtained uniformly through getCheckpointService. getCheckpointService has implemented a blocking mechanism to ensure checkpoint initialization.

File: src/core/checkpoints/index.ts [hunk 9]
```diff
@@ -161,20 +356,104 @@ async function checkGitInstallation(
...
+		)
+
+		return service
+	} catch (err) {
+		return undefined
+	}
+}
+
+// Track ongoing checkpoint saves per task to prevent duplicates
+const ongoingCheckpointSaves = new Map<string, Promise<void | CheckpointResult | undefined>>()
```
Comment (NaccOll) [inline src/core/checkpoints/index.ts]: Will starting many tasks during the life cycle cause memory problems? This seems more suitable for putting it in Task and storing it in the task history

File: src/core/checkpoints/index.ts (hunk 7 above)
Comment (playcations) [reply src/core/checkpoints/index.ts]: FileChangeManager needs a "baseline checkpoint" to start from when a task is opened or created. Checkpoints are still created initially, this is for FCO, in case one does not exist already. 

This is a conditional initialization checkpoint specifically designed to ensure FileChangeManager has a proper baseline for FCO functionality, not a postponement until first file edit.

Comment (NaccOll) [reply src/core/checkpoints/index.ts]: I understand what you mean. But the timing of the checkpoint has changed when the pr wait for merge. 

Previously, checkpoints were only saved after a file was edited, so a checkpoint needed to be saved as baseline when the checkpoint was initialized. 
But now the checkpoint will be created before the first file edit. The checkpoint before the first file edit is the baseline. 

Does this meet your requirements? Otherwise, the new code will result in two checkpoints before the first file edit, but no file modifications have been made between the two checkpoints.

File: src/core/checkpoints/index.ts (hunk 9 above)
Comment (playcations) [reply src/core/checkpoints/index.ts]: I have done what you said, makes perfect sense.

File: src/core/checkpoints/index.ts (hunk 6 above)
Comment (playcations) [reply src/core/checkpoints/index.ts]: Updated

**Comment by playcations** [comment]
Updated to handle @NaccOll's concerns and cover somethings missed.

**Comment by NaccOll** [comment]
> Updated to handle @NaccOll's concerns and cover somethings missed.

I have some new comments about saving checkpoints on initialization, can you take a look? https://github.com/RooCodeInc/Roo-Code/pull/5626#discussion_r2314384187

File: src/core/checkpoints/index.ts (hunk 7 above)
Comment (playcations) [reply src/core/checkpoints/index.ts]: You are correct, that initial check before edit should be all that is needed. What I added is probably unnecessary

Comment (playcations) [reply src/core/checkpoints/index.ts]: After some double checking the reason this is there is because without it, the baseline checkpoint gets set to the initial checkpoint when reopening tasks. I will see if I can find a better way of handling that logic.

Comment (playcations) [reply src/core/checkpoints/index.ts:171]: Here is the real issue, I had code that switched checkpoints to after a file was edited and had the initial checkpoint for that reason. I am seeing now that is not in the code anymore. I can either set that back up so all checkpoints are called after editing the files or find a new approach to this. Let me know what you think.

Comment (NaccOll) [reply src/core/checkpoints/index.ts]: This sounds a bit tricky and will have to wait for a decision from the RooCode team. The proposal to change the timing of checkpoints comes from https://github.com/RooCodeInc/Roo-Code/issues/4827

The editing and deletion of user messages awaiting review(https://github.com/RooCodeInc/Roo-Code/pull/7447) also depends on the checkpoint save timing before the file is edited.

Comment (playcations) [reply src/core/checkpoints/index.ts]: I see, so you have been messing with this quite a bit. I do think just moving checkpoints after the file is changed would solve this. Have an initial checkpoint and then go from there.

Comment (NaccOll) [reply src/core/checkpoints/index.ts]: I can resolve #7447 by simply adjusting the checkpoint references.

The main issue still stems from #4827.

Previously, checkpoints were saved after the agent edits, but this allowed for manual edits to be made between the two agent edits. If the user attempted to roll back the checkpoint after the next agent edit, they would only be able to roll back to the previous agent edit, which would also roll back the manually edited code. This is the reason for #4827. I think this is a very significant change.

I will review the code in this pull request to try to find a solution. Previously, I focused only on the impact of the changes on checkpoints.

**Comment by NaccOll** [comment]
@playcations 

Could you please tell me how to enable the FCO function? Is there something wrong with my settings? I need to test to solve the problem we mentioned.

image
image

**Comment by playcations** [comment]
@NaccOll 
What it looks like is multiple checkpoint still before the edit, it will not appear with the version you have until you make another edit. I am cleaning up and reviewing a change I am about to push that seems to work. It involves using the tool calls to update the Files Changed manager.

File: src/core/checkpoints/index.ts (hunk 7 above)
Comment (playcations) [reply src/core/checkpoints/index.ts]: pushed, let me know how that works

Comment (playcations) [reply src/core/checkpoints/index.ts]: I am off to bed and have work, I can get to this again tomorrow night around 8 pm est.

**Comment by playcations** [comment]
There is a piece that I see was not pushed (my fault) that I will get to tomorrow and that involves tracking changes to files after accept and reject have been called on them.

File: src/core/checkpoints/index.ts (hunk 7 above)
Comment (NaccOll) [reply src/core/checkpoints/index.ts]: It works! 

Can you adjust the text color of the view diff button?

image

Comment (playcations) [reply src/core/checkpoints/index.ts]: Yikes, I'll fix that.

**Comment by playcations** [comment]
I used the merge agent in Roo and now I have more. I can get to these tomorrow.

**Comment by playcations** [comment]
@roomote-agent fix the merge conflicts using merge resolver mode

**Comment by ghost** [comment]
Hi @playcations! I'm on it - I'll resolve the merge conflicts for the Files Changed Overview (FCO) PR. Let me check out the branch and analyze the conflicts now.

**Comment by ghost** [comment]
## Merge Conflict Resolution Update

I've successfully resolved all 36 merge conflicts in the PR and pushed the changes to the `fco-clean` branch. Here's what was done:

### Conflicts Resolved:
- **CHANGELOG.md**: Merged version 3.26.5 release notes from both branches
- **README files (19 total)**: Updated contributor tables across all locale versions
- **ExtensionChannel.ts**: Added new events (TaskPaused, TaskUnpaused, TaskSpawned, TaskUserMessage)
- **Type definitions**: Added missing types for the Files Changed Overview feature:
  - Added message types: `webviewReady`, `filesChangedRequest`, `filesChangedEnabled`, etc.
  - Added properties: `fileChanges`, `baseline`, `filesChangedEnabled`
  - Added methods: `getFileChangeManager()`, `ensureFileChangeManager()`
- **Fixed TypeScript compilation errors**: 
  - Added type annotations for implicit any parameters
  - Fixed Set vs Map usage for `ongoingCheckpointSaves`
  - Added missing properties to Task and ClineProvider classes

### Current Status:
The branch has been successfully pushed with all conflicts resolved and TypeScript compilation passing. However, GitHub is still showing the PR as conflicting. This may be due to:
1. GitHub needing time to update the merge status
2. The PR base branch may need to be updated

### Next Steps:
Please try refreshing the PR page or updating the base branch if the conflicts persist. All code changes have been properly merged and the branch compiles successfully.

The resolved branch is available at: `fco-clean`

**Comment by playcations** [comment]
roomote-agent fix the merge conflicts you previously mentioned using merge resolver mode

**Comment by playcations** [comment]
@roomote-agent fix the merge conflicts you previously mentioned using merge resolver mode

**Comment by hannesrudolph** [comment]
This PR has been rebased onto the latest main branch to resolve merge conflicts.

**New PR:** #7800

The rebased version preserves all the Files Changed Overview feature implementation while integrating with recent main branch changes. Please review the new PR for the updated code.

Note: The new PR has some TypeScript compilation errors that resulted from the merge conflict resolution and will need manual cleanup before merging.

**Comment by playcations** [comment]
A new PR has been created for this #7816 

A new branch was made stemming from the head of main and Files changed overview was refactored to work from there.
